README.md
# ObjectiveQuest

オブジェクト指向の学習として作ったドラ○エ風ゲーム。

## ステータス

### ユニット

- HP：体力　ゼロになると戦闘不能
- MP：マジックポイント　魔法を使う際に消費する
- ATK：攻撃力　物理ダメージに影響
- MAT：魔力　魔法ダメージに影響
- AGI：素早さ　行動順に影響

### アビリティ

- MPコスト：消費MP
- ダメージ量：与えるダメージ量（固定値）

## 開発言語

Ruby

## Change_log

機能追加・改修の履歴を残す。

各versionのソースコードは、Tagsから'vx.x'を選択することで閲覧可能。

### v1.7

#### Summary

- 回復魔法アビリティを追加
- HPが50%以下のユニットがいた場合回復使用
- 回復対象優先度はHP減少量の大きい順
- アビリティ優先度は回復量の大きい順

#### Output_sample

- `output/output_v1.7.txt`

### v1.6

#### Summary

- MPステータスを追加
- 魔法攻撃アビリティを追加
- 攻撃対象はランダム
- アビリティ優先度はダメージ量の大きい順

#### Output_sample

- `output/output_v1.6.txt`

### v1.5

#### Summary

- 素早さ（AGI）ステータスを追加
- 行動順は素早さが高い順とする

#### Output_sample

- `output/output_v1.5.txt`

### v1.4

#### Summary

- 複数人パーティーに対応
- 行動順は固定

#### Output_sample

- `output/output_v1.4.txt`

### v1.3

#### Summary

- 攻撃力タイプを物理攻撃力（ATK）と魔法攻撃力（MAT）の2種類とする
- 一定確率（確率は固定値）で魔法攻撃をする

#### Output_sample

- `output/output_v1.3.txt`

### v1.2

#### Summary

- 一定確率（確率は固定値）でミスが発生する
- 一定確率（確率は固定値）でクリティカルダメージ（係数は固定値）が発生する

#### Output_sample

乱数が使われるので同じ初期ステータス（HP,攻撃力）でも勝敗が変化する

- `output/output_v1.2_1.txt`
- `output/output_v1.2_1.txt`

### v1.1

#### Summary

- 勇者と魔王が存在する
- 勇者と魔王はそれぞれHPと攻撃力というステータス（固定値）を持つ
- 勇者と魔王は相互に攻撃を繰り返す
- 攻撃力分、相手のHPを減らす
- どちらかのHPが0以下になった場合に終了

#### Output_sample

固定値なので毎回同じ出力

- `output/output_v1.1.txt`


change_log.md
# Change_log

機能追加・改修の履歴を残す。
各versionのソースコードは、Tagsから'vx.x'を選択することで閲覧可能。

## v1.4

### Summary

- 複数人パーティーに対応
- 行動順は固定

### Output_sample

- `output/output_v1.4.txt`

## v1.3

### Summary

- 攻撃力タイプを物理攻撃力（ATK）と魔法攻撃力（MAT）の2種類とする
- 一定確率（確率は固定値）で魔法攻撃をする

### Output_sample

- `output/output_v1.3.txt`

## v1.2

### Summary

- 一定確率（確率は固定値）でミスが発生する
- 一定確率（確率は固定値）でクリティカルダメージ（係数は固定値）が発生する

### Output_sample

乱数が使われるので同じ初期ステータス（HP,攻撃力）でも勝敗が変化する

- `output/output_v1.2_1.txt`
- `output/output_v1.2_1.txt`

## v1.1

### Summary

- 勇者と魔王が存在する
- 勇者と魔王はそれぞれHPと攻撃力というステータス（固定値）を持つ
- 勇者と魔王は相互に攻撃を繰り返す
- 攻撃力分、相手のHPを減らす
- どちらかのHPが0以下になった場合に終了

### Output_sample

固定値なので毎回同じ出力

- `output/output_v1.1.txt`


main.rb
require 'pry'
require './facilitator'
require './unit'
require './ability'

hero1 = Unit.new(name: 'ゆうしゃ', max_hp: 300, max_mp: 0, atk: 50, mat: 10, agi: 50)
hero2 = Unit.new(name: '魔法使い', max_hp: 150, max_mp: 30, atk: 20, mat: 50, agi: 40)
hero3 = Unit.new(name: 'せんし', max_hp: 350, max_mp: 0, atk: 40, mat: 0, agi: 30)

enemy1 = Unit.new(name: 'スライム', max_hp: 10, max_mp: 0, atk: 10, mat: 0, agi: 10)
enemy2 = Unit.new(name: 'ドラゴン', max_hp: 250, max_mp: 50, atk: 30, mat: 90, agi: 60)
enemy3 = Unit.new(name: 'ゴーレム', max_hp: 350, max_mp: 0, atk: 80, mat: 0, agi: 20)

magic1 = Ability.new(name: 'メラ', type: Ability::TYPE[:magic_attack] , mp_cost: 8, effect_amount: 50)
magic2 = Ability.new(name: 'メラゾーマ', type: Ability::TYPE[:magic_attack] , mp_cost: 18, effect_amount: 120)
heal1 = Ability.new(name: 'ホイミ', type: Ability::TYPE[:heal] , mp_cost: 8, effect_amount: 50)
heal2 = Ability.new(name: 'ベホイミ', type: Ability::TYPE[:heal] , mp_cost: 18, effect_amount: 120)

hero2.learn([magic1, heal1, heal2])
enemy2.learn([magic1, magic2, heal1])

heros = [hero1, hero2, hero3]
enemies = [enemy1, enemy2, enemy3]

Facilitator.set(heros, enemies)
Facilitator.start


unit.rb
require 'pry'
CRITICAL_COEFFICIENT = 1.5
CRITICAL_PROBABILITY = 5
MISS_PROBABILITY = 10
HEALING_PERCENTAGE = 0.5

class Unit
  attr_reader :name, :max_hp, :atk, :mat, :agi, :abilities
  attr_accessor :hp, :mp
  def initialize(name:, max_hp:, max_mp:, atk:, mat:, agi:)
    @name = name
    @max_hp = max_hp
    @hp = @max_hp
    @max_mp = max_mp
    @mp = @max_mp
    @atk = atk
    @mat = mat
    @agi = agi
    @abilities = []
  end

  def alive?
    hp > 0
  end

  def attack(target)
    puts "#{name}の攻撃！"
    return puts 'ミス！！！' if miss?
    if magic_attack?
      magic_attack(target)
    elsif critical?
      critical_attack(target)
    else
      physical_atk(target)
    end
  end

  def show_status
    "#{name} HP:#{hp}, MP:#{mp}, 攻撃力:#{atk}, 魔法攻撃力:#{mat}, 素早さ:#{agi}"
  end

  def learn(ability_list)
    @abilities = abilities | ability_list
  end

  def heal_abilities
    abilities.select { |ability| ability.heal? && ability.mp_cost <= mp }
  end

  def damaged_amount
    max_hp - hp
  end

  def need_heal?
    alive? && hp <= max_hp * HEALING_PERCENTAGE
  end

  def heal(target)
    ability = select_healing_ability
    puts "♡回復♡#{name}は#{ability.name}を唱えた"
    heal_amount = ((target.hp + ability.effect_amount) > target.max_hp)? target.max_hp - target.hp : ability.effect_amount
    target.hp += heal_amount
    reduce_mp(ability)
    heal_message(target, heal_amount)
  end

  private

  def dead?
    !alive?
  end

  def miss?
    rand(MISS_PROBABILITY) == 1
  end

  def magic_attack?
    !magic_attack_abilities.empty? && mp > 0
  end

  def magic_attack_abilities
    abilities.select { |ability| ability.magic_attack? && ability.mp_cost <= mp }
  end

  def magic_attack(target)
    ability = select_magic_attack_ability
    puts "★魔法攻撃★#{name}は#{ability.name}を唱えた"
    damage_amount = ability.effect_amount
    reduce_mp(ability)
    damege(target, damage_amount)
  end

  def select_magic_attack_ability
    magic_attack_abilities.sort_by { |ability| ability.effect_amount }.reverse.first
  end

  def reduce_mp(ability)
    self.mp = ((mp - ability.mp_cost) < 0)? 0 : mp - ability.mp_cost
  end

  def damege(target, damage_amount)
    target.hp -= damage_amount
    dameg_message(target, damage_amount)
  end

  def dameg_message(target, damage_amount)
    puts "#{target.name}に#{damage_amount}のダメージ！"
  end

  def select_healing_ability
    heal_abilities.sort_by { |ability| ability.effect_amount }.reverse.first
  end

  def heal_message(target, heal_amount)
    puts "#{target.name}のHPを#{heal_amount}回復！"
  end

  def critical?
    rand(CRITICAL_PROBABILITY) == 1
  end

  def critical_attack(target)
    puts 'クリティカルーーー！！！'
    damage_amount = (atk * CRITICAL_COEFFICIENT).round
    damege(target, damage_amount)
  end

  def physical_atk(target)
    damage_amount = atk
    damege(target, damage_amount)
  end
end


facilitator.rb
require 'pry'
class Facilitator

  class << self
    def set(heros, enemies)
      @heros = heros
      @enemies = enemies
    end

    def start
      opening_message
      battle
    end

    private

    def heros
      @heros
    end

    def enemies
      @enemies
    end

    def opening_message
      puts 'バトル開始！'
      puts sorted_units.map(&:show_status)
      puts '================='
    end

    def all_units
      heros + enemies
    end

    def battle
      while alive_anyone?
        units = sorted_units
        units.each do |unit|
          if heal_action?(unit)
            healing_target = select_healing_target(healing_targets(unit))
            unit.heal(healing_target)
          else
            targets = heros?(unit) ? enemies : heros
            target = select_target(targets)
            unit.attack(target)
            if target.hp <= 0
              units.delete(target)
              targets.delete(target)
              puts "#{unit.name}は#{target.name}をやっつけた！"
            end
            if wipe?(targets)
              ending_message = heros?(unit) ? happy_end(unit, target) : bad_end(unit, target)
              return puts ending_message
            end
          end
          puts units.map(&:show_status)
          puts '================='
        end
      end
    end

    def alive_anyone?
      !all_units.select { |unit| unit.alive? }.empty?
    end

    def sorted_units
      all_units.sort_by { |unit| unit.agi }.reverse
    end

    def select_target(targets)
      targets.select { |target| target.alive? }.shuffle.first
    end

    def heal_action?(unit)
      !unit.heal_abilities.empty? && !healing_targets(unit).empty?
    end

    def healing_targets(unit)
      targets = heros?(unit) ? heros : enemies
      targets.select { |target| target.need_heal? == true }
    end

    def select_healing_target(targets)
      targets.sort_by { |target| target.damaged_amount }.reverse.first
    end

    def heros?(unit)
      heros.include?(unit)
    end

    def wipe?(targets)
      targets.empty?
    end

    def happy_end(offense, defense)
      "#{offense.name}たちは#{defense.name}たちを倒した！。世界に平和が訪れた。"
    end

    def bad_end(offense, defense)
      "#{defense.name}たちは倒れた・・・。世界が闇に包まれた。"
    end
  end
end


quest.rb
require 'pry'

CRITICAL_COEFFICIENT = 1.5
CRITICAL_PROBABILITY = 5
MISS_PROBABILITY = 10
MAGIC_ATTACK_PROBABILITY = 3

class Unit
  attr_reader :name, :atk, :mat, :agi
  attr_accessor :hp
  def initialize(name:, max_hp:, atk:, mat:, agi:)
    @name = name
    @max_hp = max_hp
    @hp = @max_hp
    @atk = atk
    @mat = mat
    @agi = agi
  end

  def attack(target)
    puts "#{name}の攻撃！"
    return puts 'ミス！！！' if miss?
    if magic_attack?
      puts '★魔法攻撃！'
      target.hp -= mat
      damage = mat
    elsif critical?
      puts 'クリティカルーーー！！！'
      critical_attack(target)
      damage = (atk * CRITICAL_COEFFICIENT).round
    else
      target.hp -= atk
      damage = atk
    end
    puts "#{target.name}に#{damage}のダメージ！"
  end

  def show_status
    "#{name} HP:#{hp}, 攻撃力:#{atk}, 魔法攻撃力:#{mat}, 素早さ:#{agi}"
  end

  private

  def critical?
    rand(CRITICAL_PROBABILITY) == 1
  end

  def miss?
    rand(MISS_PROBABILITY) == 1
  end

  def magic_attack?
    rand(MAGIC_ATTACK_PROBABILITY) == 1
  end

  def critical_attack(target)
    target.hp -= (atk * CRITICAL_COEFFICIENT).round
  end
end

class Facilitator
  attr_reader :heros, :enemies
  def initialize(heros, enemies)
    @heros = set_units(heros)
    @enemies = set_units(enemies)
  end

  def start
    opening_message
    battle
  end

  private

  def set_units(units_infos)
    units = []
    units_infos.each do |info|
      units << Unit.new(name: info[:name], max_hp: info[:max_hp], atk: info[:atk], mat: info[:mat], agi: info[:agi])
    end
    units
  end

  def opening_message
    puts 'バトル開始！'
    puts sorted_units.map(&:show_status)
    puts '================='
  end

  def all_units
    heros + enemies
  end

  def battle
    while alive_anyone?
      units = sorted_units
      units.each do |unit|
        targets = heros?(unit) ? enemies : heros
        target = select_target(targets)
        unit.attack(target)
        if target.hp <= 0
          units.delete(target)
          targets.delete(target)
          puts "#{unit.name}は#{target.name}をやっつけた！"
        end
        if wipe?(targets)
          ending_message = heros?(unit) ? happy_end(unit, target) : bad_end(unit, target)
          return puts ending_message
        end
        puts units.map(&:show_status)
        puts '================='
      end
    end
  end

  def alive_anyone?
    flg = false
    all_units.each {|unit| break flg = true if unit.hp > 0 }
    flg
  end

  def sorted_units
    all_units.sort_by { |unit| unit.agi }.reverse
  end

  def select_target(targets)
    target = targets.shuffle.first
    if target.hp <= 0
      targets.delete(target)
      select_target(targets)
    end
    target
  end

  def heros?(unit)
    heros.include?(unit)
  end

  def wipe?(defenses)
    defenses.empty?
  end

  def happy_end(offense, defense)
    "#{offense.name}たちは#{defense.name}たちを倒した！。世界に平和が訪れた。"
  end

  def bad_end(offense, defense)
    "#{defense.name}たちは倒れた・・・。世界が闇に包まれた。"
  end

  def change_turn(offenses, defenses)
    [defenses, offenses]
  end
end

HERO1 = { name: 'ゆうしゃ', max_hp: 200, atk: 50, mat: 10, agi: 50 }
HERO2 = { name: '魔法使い', max_hp: 100, atk: 20, mat: 50, agi: 40 }
HERO3 = { name: 'せんし', max_hp: 400, atk: 40, mat: 0, agi: 30 }

ENEMY1 = { name: 'スライム', max_hp: 10, atk: 10, mat: 0, agi: 10 }
ENEMY2 = { name: 'ドラゴン', max_hp: 250, atk: 30, mat: 90, agi: 60 }
ENEMY3 = { name: 'ゴーレム', max_hp: 400, atk: 70, mat: 0, agi: 20 }

heros = [HERO1, HERO2, HERO3]
enemies = [ENEMY1, ENEMY2, ENEMY3]

facilitator = Facilitator.new(heros, enemies)
facilitator.start
